capability/keeper/keeper.go

```diff
@@ -0,0 +1,494 @@
+	// to claim capabilities they receive and retrieve capabilities which they own
+	// by name, in addition to creating new capabilities & authenticating capabilities
+	// passed by other modules.
+	ScopedKeeper struct {
+		cdc      codec.BinaryCodec
+		storeKey storetypes.StoreKey
+		memKey   storetypes.StoreKey
+		capMap   map[uint64]*types.Capability
+		module   string
+	}
+)
+
+// NewKeeper constructs a new CapabilityKeeper instance and initializes maps
+// for capability map and scopedModules map.
+func NewKeeper(cdc codec.BinaryCodec, storeKey, memKey storetypes.StoreKey) *Keeper {
+	return &Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
+		capMap:        make(map[uint64]*types.Capability),
+		scopedModules: make(map[string]struct{}),
+		sealed:        false,
+	}
+}
+
+// ScopeToModule attempts to create and return a ScopedKeeper for a given module
+// by name. It will panic if the keeper is already sealed or if the module name
+// already has a ScopedKeeper.
+func (k *Keeper) ScopeToModule(moduleName string) ScopedKeeper {
+	if k.sealed {
+		panic("cannot scope to module via a sealed capability keeper")
+	}
+	if strings.TrimSpace(moduleName) == "" {
+		panic("cannot scope to an empty module name")
+	}
+
+	if _, ok := k.scopedModules[moduleName]; ok {
+		panic(fmt.Sprintf("cannot create multiple scoped keepers for the same module name: %s", moduleName))
+	}
+
+	k.scopedModules[moduleName] = struct{}{}
+
+	return ScopedKeeper{
+		cdc:      k.cdc,
+		storeKey: k.storeKey,
+		memKey:   k.memKey,
+		capMap:   k.capMap,
+		module:   moduleName,
+	}
+}
+
+// Seal seals the keeper to prevent further modules from creating a scoped keeper.
+// Seal may be called during app initialization for applications that do not wish to create scoped keepers dynamically.
+func (k *Keeper) Seal() {
+	if k.sealed {
+		panic("cannot initialize and seal an already sealed capability keeper")
+	}
+
+	k.sealed = true
+}
+
+// IsSealed returns if the keeper is sealed.
+func (k *Keeper) IsSealed() bool {
+	return k.sealed
+}
+
+// InitMemStore will assure that the module store is a memory store (it will panic if it's not)
+// and willl initialize it. The function is safe to be called multiple times.
+// InitMemStore must be called every time the app starts before the keeper is used (so
+// `BeginBlock` or `InitChain` - whichever is first). We need access to the store so we
+// can't initialize it in a constructor.
+func (k *Keeper) InitMemStore(ctx sdk.Context) {
+	memStore := ctx.KVStore(k.memKey)
+	memStoreType := memStore.GetStoreType()
+
+	if memStoreType != storetypes.StoreTypeMemory {
+		panic(fmt.Sprintf("invalid memory store type; got %s, expected: %s", memStoreType, storetypes.StoreTypeMemory))
+	}
+
+	// create context with no block gas meter to ensure we do not consume gas during local initialization logic.
+	noGasCtx := ctx.WithBlockGasMeter(storetypes.NewInfiniteGasMeter())
```

Comment:
this is one pr we should adopt as well https://github.com/cosmos/cosmos-sdk/pull/15030